File: src/go/array/spiralOrder.go
package array

// 54. 螺旋矩阵
func spiralOrder(matrix [][]int) []int {
	left, top := 0, 0
	right := len(matrix[0])
	down := len(matrix)
	var num = right * down
	right--
	down--
	var ans []int
	for num > 0 {
		for i := left; i <= right && num > 0; i++ {
			ans = append(ans, matrix[top][i])
			num--
		}
		top++
		for i := top; i <= down && num > 0; i++ {
			ans = append(ans, matrix[i][right])
			num--
		}
		right--
		for i := right; i >= left && num > 0; i-- {
			ans = append(ans, matrix[down][i])
			num--
		}
		down--
		for i := down; i >= top && num > 0; i-- {
			ans = append(ans, matrix[i][left])
			num--
		}
		left++
	}
	return ans
}

File: feishu/SpiralOrder.java
import java.util.ArrayList;
import java.util.List;

/**
 * @author Zexho
 * @date 2022/3/28 10:51 PM
 */
public class SpiralOrder {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> ans = new ArrayList<>();
        process(matrix, 0, 0, matrix[0].length - 1, matrix.length - 1, ans);
        return ans;
    }

    private void process(int[][] arr, int left, int up, int right, int down, List<Integer> ans) {
        for (int i = left; i <= right; i++) {
            ans.add(arr[up][i]);
        }
        for (int i = up + 1; i <= down; i++) {
            ans.add(arr[i][right]);
        }
        if (left < right && down > up) {
            for (int i = right - 1; i >= left; i--) {
                ans.add(arr[down][i]);
            }
            for (int i = down - 1; i > up; i--) {
                ans.add(arr[i][left]);
            }
        }
        if (left < right - 1 && up < down - 1) {
            process(arr, left + 1, up + 1, right - 1, down - 1, ans);
        }
    }

}
